src/zarr/core/metadata/v3.py

```diff
@@ -359,9 +394,18 @@ def parse_fill_value(
     ...
 
 
+def default_fill_value(dtype: DataType) -> str | bytes | np.generic:
+    if dtype == DataType.string:
+        return ""
+    elif dtype == DataType.bytes:
+        return b""
+    else:
+        return dtype.to_numpy().type(0)
+
+
 def parse_fill_value(
-    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool | None,
-    dtype: BOOL_DTYPE | INTEGER_DTYPE | FLOAT_DTYPE | COMPLEX_DTYPE | np.dtype[Any],
+    fill_value: complex | str | bytes | np.generic | Sequence[Any] | bool,
+    dtype: DataType,
```

Comment:
I have changed this function to accept a Zarr v3 `DataType` rather than a numpy dtype.
All the tests are working. However, mypy is not happy.

```
src/zarr/core/metadata/v3.py:219: error: No overload variant of "parse_fill_value" matches argument types "Any", "DataType"  [call-overload]
src/zarr/core/metadata/v3.py:219: note: Possible overload variants:
src/zarr/core/metadata/v3.py:219: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: BoolDType) -> bool
src/zarr/core/metadata/v3.py:219: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool, dtype: Int8DType | Int16DType | Int32DType | Int64DType | UInt8DType | UInt16DType | UInt32DType | UInt64DType) -> signedinteger[_8Bit] | signedinteger[_16Bit] | signedinteger[_32Bit] | signedinteger[_64Bit] | unsignedinteger[_8Bit] | unsignedinteger[_16Bit] | unsignedinteger[_32Bit] | unsignedinteger[_64Bit]
src/zarr/core/metadata/v3.py:219: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool | None, dtype: Float16DType | Float32DType | Float64DType) -> floating[_16Bit] | floating[_32Bit] | floating[_64Bit]
src/zarr/core/metadata/v3.py:219: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool | None, dtype: Complex64DType | Complex128DType) -> complexfloating[_32Bit, _32Bit] | complexfloating[_64Bit, _64Bit]
src/zarr/core/metadata/v3.py:219: note:     def parse_fill_value(fill_value: complex | str | bytes | generic | Sequence[Any] | bool | None, dtype: dtype[Any]) -> Any
src/zarr/core/metadata/v3.py:354: error: An overloaded function outside a stub file must have an implementation  [no-overload-impl]
src/zarr/core/metadata/v3.py:403: error: Returning Any from function declared to return "str | bytes | generic"  [no-any-return]
src/zarr/core/metadata/v3.py:406: error: Name "parse_fill_value" already defined on line 354  [no-redef]
src/zarr/core/metadata/v3.py:442: error: Incompatible types in assignment (expression has type "Complex64DType | Complex128DType", variable has type "DataType")  [assignment]
src/zarr/core/array.py:321: error: Argument 1 to "get_default_array_bytes_codec" has incompatible type "dtype[Any] | type[Any] | _SupportsDType[dtype[Any]] | str | tuple[Any, int] | tuple[Any, SupportsIndex | Sequence[SupportsIndex]] | list[Any] | _DTypeDict | tuple[Any, Any] | None"; expected "dtype[Any]"  [arg-type]
```

Seeking advice and what is the goal of the this `@overload` stuff and how it might be refactored to work with `DataType` rather than numpy dtypes.